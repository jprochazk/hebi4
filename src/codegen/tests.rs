use std::{fs::read_to_string, path::Path};

use super::emit;
use crate::{codegen::EmitOptions, parser::parse, token::tokenize};

fn parse_options(options: &str) -> EmitOptions {
    let mut o = EmitOptions::empty();

    for opt in options.split(',') {
        match opt {
            "dce" => o.dead_code_elimination = true,
            _ => {}
        }
    }

    o
}

#[glob_test::glob("../../tests/inputs/codegen/*.hi")]
fn emitter(path: &Path) {
    let input = read_to_string(path).unwrap();
    let input = input.trim();
    let options = if input.starts_with("//") {
        let first_line = input.lines().next().unwrap().trim_start_matches("//");
        parse_options(first_line)
    } else {
        Default::default()
    };

    let tokens = tokenize(input);
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(input));
        }
    };
    let (snapshot, failure) = match emit("test".into(), &ast, options) {
        Ok(m) => (format!("SOURCE\n{input}\n\nOK\n{}", m.disasm(input)), false),
        Err(err) => (
            format!("SOURCE\n{input}\n\nERROR\n{}", err.render(input)),
            true,
        ),
    };

    #[cfg(not(miri))]
    {
        insta::assert_snapshot!(snapshot);
        let _ = failure;
    }

    #[cfg(miri)]
    {
        if failure {
            panic!("{snapshot}");
        } else {
            eprintln!("{snapshot}");
        }
    }
}

// NOTE: leave this here, it's handy for debugging:
// 1. uncomment this test
// 2. filter for it (`cargo test -- _temp`),
// 3. set a breakpoint somewhere in codegen

// #[test]
// fn _temp() {
//     emitter(std::path::Path::new("tests/inputs/run/arithmetic_add.hi"));
//     panic!();
// }
