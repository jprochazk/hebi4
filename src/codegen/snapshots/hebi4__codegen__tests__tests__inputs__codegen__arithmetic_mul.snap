---
source: src/codegen/tests.rs
expression: snapshot
---
SOURCE
let result = 0
let a = 1
let b = 1

result = a * b
result = 1 * b
result = a * 1

OK
.fn @main
  params 0
  stack 4
  .locals
    result = r1
    a = r2
    b = r3
  .code
  0  lsmi r1, 0
  1  lsmi r2, 1
  2  lsmi r3, 1
  3  mul r1, r2, r3
  4  mul r1, 1, r3
  5  mul r1, r2, 1
  6  lnil r0
  7  ret
